core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2019-2022 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlinx.datetime.internal.clampToInt
+import kotlinx.datetime.internal.safeAdd
+import kotlinx.datetime.internal.safeMultiplyOrClamp
+import kotlin.random.Random
+import kotlin.random.nextLong
+
+private class LocalDateProgressionIterator(private val iterator: LongIterator) : Iterator<LocalDate> {
+    override fun hasNext(): Boolean = iterator.hasNext()
+    override fun next(): LocalDate = LocalDate.fromEpochDays(iterator.next())
+}
+
+/**
+ * A progression of values of type [LocalDate].
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.reversedProgression
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.firstAndLast
+ */
+public open class LocalDateProgression
+internal constructor(internal val longProgression: LongProgression) : Collection<LocalDate> {
+
+    internal constructor(
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Long
+    ) : this(LongProgression.fromClosedRange(start.toEpochDaysLong(), endInclusive.toEpochDaysLong(), step))
+
+    /**
+     * Returns the first [LocalDate] of the progression
+     */
+    public val first: LocalDate = LocalDate.fromEpochDays(longProgression.first)
+
+    /**
+     * Returns the last [LocalDate] of the progression
+     */
+    public val last: LocalDate = LocalDate.fromEpochDays(longProgression.last)
+
+    /**
+     * Returns an [Iterator] that traverses the progression from [first] to [last]
+     */
+    override fun iterator(): Iterator<LocalDate> = LocalDateProgressionIterator(longProgression.iterator())
+
+    /**
+     * Returns true iff the progression contains no values.
+     * i.e. [first] < [last] if step is positive, or [first] > [last] if step is negative.
+     */
+    public override fun isEmpty(): Boolean = longProgression.isEmpty()
+
+    /**
+     * Returns a string representation of the progression.
+     * Uses the range operator notation if the progression is increasing, and `downTo` if it is decreasing.
+     * The step is referenced in days.
+     */
+    override fun toString(): String = if (longProgression.step > 0) "$first..$last step ${longProgression.step}D" else "$first downTo $last step ${longProgression.step}D"
+
+    /**
+     * Returns the number of dates in the progression.
+     * Returns [Int.MAX_VALUE] if the number of dates overflows [Int]
+     */
+    override val size: Int
+        get() = longProgression.size
+
+    /**
+     * Returns true iff every element in [elements] is a member of the progression.
+     */
+    override fun containsAll(elements: Collection<LocalDate>): Boolean = elements.all(::contains)
```

Comment:
Is this a deeper issue than at first appears? The signature of the methods implies no such unsafety, however after digging into it I can see that the methods in `AbstractCollection` are annotated with `@UnsafeVariance`, while the methods in `UByteArray` contain explicit type checks. I suppose it's a concession to Java compatibility where the `contains` methods take instances of `Object` instead of a typed parameter?

In any event, I'll copy the type check logic from `UByteArray` and add the test.